pytket/extensions/cutensornet/tnstate/ttn.py

```diff
@@ -0,0 +1,845 @@
+            unsafe: If True, the final state will be different than the starting one.
+                Specifically, the information in the returned bond tensor at ``center``
+                is removed from the TTN. It is expected that the caller will reintroduce
+                the bond tensor after some processing (e.g. after SVD truncation).
+
+        Returns:
+            The bond tensor created at ``center`` when canonicalisation is complete.
+            Applying SVD to this tensor yields the global SVD of the TTN.
+
+        Raises:
+            ValueError: If the ``center`` is ``tuple()``.
+        """
+        self._logger.debug(f"Canonicalising to {str(center)}")
+        options = {"handle": self._lib.handle, "device_id": self._lib.device_id}
+
+        if isinstance(center, Qubit):
+            target_path = self.qubit_position[center][0]
+            assert not unsafe  # Unsafe disallowed when ``center`` is a qubit
+        elif center == ():
+            raise ValueError("There is no bond at path ().")
+        else:
+            target_path = center
+
+        # Separate nodes to be canonicalised towards children from those towards parent
+        towards_child = []
+        towards_parent = []
+        for path in self.nodes.keys():
+            # Nodes towards children are closer to the root and coincide in the path
+            if len(path) < len(target_path) and all(
+                path[l] == target_path[l] for l in range(len(path))
+            ):
+                towards_child.append(path)
+            # If the center is a physical bond (qubit), its node is skipped
+            elif path == target_path and isinstance(center, Qubit):
+                continue
+            # All other nodes are canonicalised towards their parent
+            else:
+                towards_parent.append(path)
+        # Sanity checks
+        assert len(towards_child) != 0
+        assert len(towards_parent) != 0
+
+        # Glossary of bond IDs
+        # chr(x) -> bond of the x-th qubit in the node (if it is a leaf)
+        # l -> left child bond of the TTN node
+        # r -> right child bond of the TTN node
+        # p -> parent bond of the TTN node
+        # s -> bond between Q and R after decomposition
+
+        # Canonicalise nodes towards parent, start from the furthest away from root
+        for path in sorted(towards_parent, key=len, reverse=True):
+            self._logger.debug(f"Canonicalising node at {path} towards parent.")
+
+            # If already in desired canonical form, do nothing
+            if self.nodes[path].canonical_form == DirTTN.PARENT:
+                self._logger.debug("Skipping, already in canonical form.")
+                continue
+
+            # Otherwise, apply QR decomposition
+            if self.nodes[path].is_leaf:
+                n_qbonds = len(self.nodes[path].tensor.shape) - 1  # Num of qubit bonds
+                q_bonds = "".join(chr(x) for x in range(n_qbonds))

```

Comment:
What if `chr(x)` becomes one of `{l,p,r,s}`? I suppose you assume that would be too much for one node, but it could be a source of a bug if not?

Reply:
Yeah, that's right. It is an extremely unlikely bug though, since leaves with more than 20 qubits are unreasonable to simulate. Since the first uppercase letter is at `chr(65)=='A'` and the first lowercase letter is at `chr(97)=='a'`, this should never be an issue.

In any case, I'm happy to impose in the `leaf_size` parameter in `Config` that `leaf_size < 65`, that'll make sure this collision can never happen.

Reply:
Ohh I see, I thought alphabet starts with `0` - haven't checked, sorry!
Still, I think it is reasonable to have this safety measure to be future-proof.